lightning/src/offers/mod.rs

```diff
@@ -0,0 +1,579 @@
+/// currency.
+#[derive(Clone, Debug)]
+pub enum Amount {
+	/// An amount of bitcoin.
+	Bitcoin {
+		/// The amount in millisatoshi.
+		amount_msats: u64,
+	},
+	/// An amount of currency specified using ISO 4712.
+	Currency {
+		/// An ISO 4712 three-letter currency code (e.g., USD).
+		iso4217_code: String,
+		/// The amount in the currency unit adjusted by the ISO 4712 exponent (e.g., USD cents).
+		amount: u64,
+	},
+}
+
+const ISO4217_CODE_LEN: usize = 3;
+
+///
+#[derive(Clone, Debug)]
+pub enum Destination {
+	///
+	NodeId(XOnlyPublicKey),
+	///
+	Paths(Vec<BlindedPath>),
+}
+
+impl Destination {
+	fn node_id(&self) -> XOnlyPublicKey {
+		match self {
+			Destination::NodeId(node_id) => *node_id,
+			Destination::Paths(paths) => unimplemented!(),
+		}
+	}
+}
+
+///
+#[derive(Clone, Debug)]
+pub struct SendInvoice {
+	refund_for: Option<PaymentHash>,
+}
+
+/// An `offer` TLV stream without any semantic checks, apart from any checks performed when parsing
+/// the underlying types.
+#[derive(Debug)]
+struct OfferTlvStream {
+	chains: Option<WithoutLength<Vec<BlockHash>>>,
+	currency: Option<WithoutLength<String>>,
+	amount: Option<HighZeroBytesDroppedVarInt<u64>>,
+	description: Option<WithoutLength<String>>,
+	features: Option<OfferFeatures>,
+	absolute_expiry: Option<HighZeroBytesDroppedVarInt<u64>>,
+	paths: Option<WithoutLength<Vec<BlindedPath>>>,
+	issuer: Option<WithoutLength<String>>,
+	quantity_min: Option<HighZeroBytesDroppedVarInt<u64>>,
+	quantity_max: Option<HighZeroBytesDroppedVarInt<u64>>,
+	recurrence: Option<Recurrence>,
+	node_id: Option<XOnlyPublicKey>,
+	send_invoice: Option<()>,
+	refund_for: Option<PaymentHash>,
+	signature: Option<Signature>,
+}
+
+impl_writeable_tlv_stream!(OfferTlvStream, {
+	(2, chains, option),
+	(6, currency, option),
+	(8, amount, option),
+	(10, description, option),
+	(12, features, option),
+	(14, absolute_expiry, option),
+	(16, paths, option),
+	(20, issuer, option),
+	(22, quantity_min, option),
+	(24, quantity_max, option),
+	(26, recurrence, option),
+	(30, node_id, option),
+	(34, refund_for, option),
+	(54, send_invoice, option),
+	(240, signature, option),
+});
+
+#[derive(Clone, Debug)]
+///
+pub struct BlindedPath {
+	blinding: PublicKey,
+	path: WithLength<Vec<OnionMessagePath>, u8>,
+}
+
+impl_writeable!(BlindedPath, { blinding, path });
+
+#[derive(Clone, Debug)]
+struct OnionMessagePath {
+	node_id: PublicKey,
+	encrypted_recipient_data: Vec<u8>,
+}
+
+impl_writeable!(OnionMessagePath, { node_id, encrypted_recipient_data });
+
+// TODO: Remove once BOLT 12 test vectors are updated.
+#[derive(Clone, Debug)]
+struct Recurrence {
+	time_unit: u8,
+	period: HighZeroBytesDroppedVarInt<u32>,
+}
+
+impl_writeable!(Recurrence, { time_unit, period });
+
+/// An offer parsed from a bech32-encoded string as a TLV stream and the corresponding bytes. The
+/// latter is used for signature verification.
+struct ParsedOffer(OfferTlvStream, Vec<u8>);
```

Comment:
> I'm really confused here now - so when we're parsing we first read some TLV fields from bytes, then we copy the bytes into this, with the TLV fields moved into it,

The bytes are not copied -- they are moved just as the TLV fields are.

> then we do some analysis of the TLV fields and copy them into a higher-level struct?

I believe they are moved into the higher level struct since `OfferTlvStream` is a value type -- I'd get compiler errors if I tried to reuse any part of the destructured `OfferTlvStream` after it had been moved.

> I'd prefer we just copy the 10 lines of TLV-read code into the top of the `Offer` read method and call it a day, all the copying around of things just seems to add more code and boilerplate and I'm still not sure how it makes anything more readable. Can you help me understand why all the boilerplate here is a simpler design?

I'm not sure how that is materially any different. We'd still need to declare variables prior to those lines, at least some with explicit types. And then repeat the types in the write code, ensuring the vector and integer types match and are serialized correctly rather than relying on the compiler to enforce it.

Additionally, that would require changing the return type to `ParseError` (in order to wrap `DecodeError`) even though all the other errors are `SemanticError`. And then all the error returns would need to be wrapped or littered with `into()`. It would also be possible to return `ParseError` variants which shouldn't be possible at that level.

I pushed a version to show what this looks like. I personally find it much less readable in addition to the other problems that it introduces as mentioned above. Moreover, the `OfferTlvStream` version is much more compact and self-describing, IMHO. I don't see how it introduces boilerplate and more code compared to the alternative.

Reply:
Discussed this offline yesterday. I pushed a new HEAD commit after reverting the non-`OfferTlvStream` version. The new version avoids any clones when writing. It also hides the `WithoutLength` and `HighZeroBytesDroppedVarInt` types from the `OfferTlvStream` struct definition and consolidates the field and TLV type definitions.

The conversions still need to be aware of the wrapper types, though. Maybe there's a neat way to avoid that with type coercions?

Reply:
Added some `From` implementations to hide all the `WithoutLength` and `HighZeroBytesDroppedVarInt` wrappers.